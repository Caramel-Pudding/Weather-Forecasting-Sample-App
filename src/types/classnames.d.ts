declare module "classnames" {
  function classNames(...classes: string[]): string;
  export default classNames;
}
