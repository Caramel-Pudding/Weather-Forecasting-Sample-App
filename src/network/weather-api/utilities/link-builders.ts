import { apiKeyStoredInACompleteSecurity } from "@/consts/mocked";
import { weatherAPIBasicRoute } from "../consts/routes";

export const buildWeatherAPIRequestRoute = (query: string): string =>
  `${weatherAPIBasicRoute}/data/2.5/forecast?q=${query}&appid=${apiKeyStoredInACompleteSecurity}`;
