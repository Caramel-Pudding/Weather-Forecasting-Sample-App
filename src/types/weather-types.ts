export interface City {
  id: number;
  country: string;
  name: string;
}

export interface WeatherData {
  city: City;
}
