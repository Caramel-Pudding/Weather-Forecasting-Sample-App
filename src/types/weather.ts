export enum PartOfTheDay {
  n = "night",
  d = "day",
}

export enum WeatherType {
  Clear = "Clear",
  Rain = "Rain",
  Snow = "Snow",
  Extreme = "Extreme",
}

export interface City {
  id: number;
  country: string;
  name: string;
}

export interface WeatherListItem {
  clouds: {
    all: number;
  };
  dt: number;
  dt_txt: string;
  main: {
    grnd_level: number;
    humidity: number;
    pressure: number;
    sea_level: number;
    temp: number;
    temp_kf: number;
    temp_max: number;
    temp_min: number;
  };
  rain?: {
    "3h"?: number;
  };
  snow?: {
    "3h"?: number;
  };
  sys: {
    pod: PartOfTheDay;
  };
  weather: {
    description: string;
    icon: string;
    id: number;
    main: WeatherType;
  }[];
  wind: {
    deg: number;
    speed: number;
  };
}

export interface WeatherData {
  city: City;
  cnt: number;
  cod: string;
  message: number;
  list: WeatherListItem[];
}
