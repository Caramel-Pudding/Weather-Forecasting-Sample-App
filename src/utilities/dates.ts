import { DayOfTheWeek, Month } from "@/consts/dates";

export const convertUnixTimestampToDate = (unixTimestamp: number): Date =>
  new Date(unixTimestamp * 1000);

export const getDayAsTextFromTimestamp = (timestamp: Date): string =>
  DayOfTheWeek[timestamp.getDay()];

export const getMonthAsTextFromTimestamp = (timestamp: Date): string =>
  Month[timestamp.getMonth()];

export const getTimeAsTextFromTimestamp = (timestamp: Date): string => {
  return `${timestamp.getHours()}:00`;
};
