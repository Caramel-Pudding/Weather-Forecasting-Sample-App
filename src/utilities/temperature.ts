export const convertKelvinToCelsius = (kelvin: number): number =>
  Math.round(kelvin - 273.15);
