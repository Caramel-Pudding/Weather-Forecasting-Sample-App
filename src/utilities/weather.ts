import { WeatherListItem } from "@/types/weather";

export const getWeatherListForToday = (
  weatherList: WeatherListItem[]
): WeatherListItem[] => {
  if (!weatherList || weatherList.length === 0) {
    return [];
  }
  const todayDate = new Date(weatherList[0].dt_txt).getDate();
  return weatherList.filter(
    (listItem) => new Date(listItem.dt_txt).getDate() === todayDate
  );
};

export const getTimeStringFromWeatherTimestamp = (
  weatherTimestamp: string
): string => {
  const timestamp = new Date(weatherTimestamp);
  return `${timestamp.getHours()}:00`;
};
