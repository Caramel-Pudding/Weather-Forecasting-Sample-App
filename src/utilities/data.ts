import { WeatherListItem } from "@/types/weather";
import { convertUnixTimestampToDate } from "./dates";

export const getWeatherListForToday = (
  weatherList: WeatherListItem[]
): WeatherListItem[] => {
  if (!weatherList || weatherList.length === 0) {
    return [];
  }

  const todayDate = convertUnixTimestampToDate(weatherList[0].dt).getDate();
  return weatherList.filter(
    (listItem) =>
      convertUnixTimestampToDate(listItem.dt).getDate() === todayDate
  );
};
