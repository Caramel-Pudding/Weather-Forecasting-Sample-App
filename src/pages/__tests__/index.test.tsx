import React from "react";
import { render, cleanup, fireEvent } from "@/tests/test-utils";

import { WeatherData } from "@/types/weather";
import { weatherDataStub } from "@/tests/stubs";
import {
  errorMainTestId,
  homeMainTestId,
  currentTemperatureTestId,
} from "@/consts/test-ids";
import { convertKelvinToCelsius } from "@/utilities/temperature";
import Home from "..";

describe("The Home Page Component", () => {
  afterEach(cleanup);

  it("Should render properly if there's proper data", () => {
    // * #TEST: ARRANGE
    const { getByTestId, queryAllByRole } = render(
      <Home weatherData={weatherDataStub as WeatherData} />,
      {}
    );
    const main = getByTestId(homeMainTestId);
    const currentTemperature = getByTestId(currentTemperatureTestId);
    const allListButtons = queryAllByRole("button") as HTMLElement[];
    // * #TEST: ASSERT
    expect(main).toBeInTheDocument();
    expect(currentTemperature).toHaveTextContent(
      String(convertKelvinToCelsius(weatherDataStub.list[0].main.temp))
    );
    expect(allListButtons).toHaveLength(3);
    expect(allListButtons[0]).toHaveClass("containerActive");
    expect(main).toMatchSnapshot();
  });

  it("Should show 404 if there's no data", () => {
    // * #TEST: ARRANGE
    const { getByTestId } = render(
      <Home weatherData={null as unknown as WeatherData} />,
      {}
    );
    const main = getByTestId(errorMainTestId);
    // * #TEST: ASSERT
    expect(main).toBeInTheDocument();
    expect(main).toMatchSnapshot();
  });

  it("Should show 404 if there's invalid data", () => {
    // * #TEST: ARRANGE
    const { getByTestId } = render(
      <Home weatherData={{} as WeatherData} />,
      {}
    );
    const main = getByTestId(errorMainTestId);
    // * #TEST: ASSERT
    expect(main).toBeInTheDocument();
    expect(main).toMatchSnapshot();
  });

  it("Should properly change selected on click on other weather item", () => {
    // * #TEST: ARRANGE
    const { container, queryAllByRole, getByTestId } = render(
      <Home weatherData={weatherDataStub as WeatherData} />,
      {}
    );
    const allListButtons = queryAllByRole("button") as HTMLElement[];
    const currentTemperature = getByTestId(currentTemperatureTestId);
    // * #TEST: ASSERT
    expect(allListButtons).toHaveLength(3);
    expect(currentTemperature).toHaveTextContent(
      String(convertKelvinToCelsius(weatherDataStub.list[0].main.temp))
    );
    expect(allListButtons[0]).toHaveClass("containerActive");
    expect(allListButtons[1]).toHaveClass("containerInactive");
    expect(container).toMatchSnapshot();
    // * #TEST: ACT
    fireEvent.click(allListButtons[1]);
    // * #TEST: ASSERT
    expect(currentTemperature).toHaveTextContent(
      String(convertKelvinToCelsius(weatherDataStub.list[1].main.temp))
    );
    expect(allListButtons[0]).toHaveClass("containerInactive");
    expect(allListButtons[1]).toHaveClass("containerActive");
    expect(container).toMatchSnapshot();
  });

  it("Should not change selected on click on the same weather item", () => {
    // * #TEST: ARRANGE
    const { container, queryAllByRole, getByTestId } = render(
      <Home weatherData={weatherDataStub as WeatherData} />,
      {}
    );
    const allListButtons = queryAllByRole("button") as HTMLElement[];
    const currentTemperature = getByTestId(currentTemperatureTestId);
    // * #TEST: ASSERT
    expect(allListButtons).toHaveLength(3);
    expect(currentTemperature).toHaveTextContent(
      String(convertKelvinToCelsius(weatherDataStub.list[0].main.temp))
    );
    expect(allListButtons[0]).toHaveClass("containerActive");
    expect(container).toMatchSnapshot();
    // * #TEST: ACT
    fireEvent.click(allListButtons[0]);
    // * #TEST: ASSERT
    expect(currentTemperature).toHaveTextContent(
      String(convertKelvinToCelsius(weatherDataStub.list[0].main.temp))
    );
    expect(allListButtons[0]).toHaveClass("containerActive");
    expect(container).toMatchSnapshot();
  });
});
