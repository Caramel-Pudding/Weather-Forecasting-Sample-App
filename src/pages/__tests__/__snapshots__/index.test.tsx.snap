// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Home Page Component Should not change selected on click on the same weather item 1`] = `
<div>
  <main
    class="main"
    data-testid="HOME-MAIN-TEST-ID"
  >
    <article
      class="container"
    >
      <article
        class="container"
      >
        <section
          class="imageWrapper"
        >
          <img
            alt="Clear"
            src="icons/weather-sun.svg"
          />
        </section>
        <section
          class="weatherWrapper"
          data-testid="CURRENT-TEMPERATURE-TEST-ID"
        >
          <div
            class="additionalText weatherHeader"
          >
            <span>
              Clear
            </span>
            <span>
              14° / 8°
            </span>
          </div>
          <div
            class="mainText mainTemperature"
          >
            14°
          </div>
        </section>
      </article>
      <article
        class="container"
      >
        <section>
          <div
            class="additionalText"
          >
            Altstadt
          </div>
          <section
            class="mainText date"
          >
            <div>
              Friday
            </div>
            <div>
              16. February
            </div>
          </section>
        </section>
      </article>
    </article>
    <section
      class="container"
    >
      <ol
        class="weatherItemsRow"
      >
        <li
          class="listItem"
        >
          <button
            class="container containerActive"
            type="button"
          >
            <div
              class="additionalText"
            >
              15:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              14°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              18:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              13°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              21:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              4°
            </div>
          </button>
        </li>
      </ol>
    </section>
  </main>
</div>
`;

exports[`The Home Page Component Should not change selected on click on the same weather item 2`] = `
<div>
  <main
    class="main"
    data-testid="HOME-MAIN-TEST-ID"
  >
    <article
      class="container"
    >
      <article
        class="container"
      >
        <section
          class="imageWrapper"
        >
          <img
            alt="Clear"
            src="icons/weather-sun.svg"
          />
        </section>
        <section
          class="weatherWrapper"
          data-testid="CURRENT-TEMPERATURE-TEST-ID"
        >
          <div
            class="additionalText weatherHeader"
          >
            <span>
              Clear
            </span>
            <span>
              14° / 8°
            </span>
          </div>
          <div
            class="mainText mainTemperature"
          >
            14°
          </div>
        </section>
      </article>
      <article
        class="container"
      >
        <section>
          <div
            class="additionalText"
          >
            Altstadt
          </div>
          <section
            class="mainText date"
          >
            <div>
              Friday
            </div>
            <div>
              16. February
            </div>
          </section>
        </section>
      </article>
    </article>
    <section
      class="container"
    >
      <ol
        class="weatherItemsRow"
      >
        <li
          class="listItem"
        >
          <button
            class="container containerActive"
            type="button"
          >
            <div
              class="additionalText"
            >
              15:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              14°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              18:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              13°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              21:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              4°
            </div>
          </button>
        </li>
      </ol>
    </section>
  </main>
</div>
`;

exports[`The Home Page Component Should properly change selected on click on other weather item 1`] = `
<div>
  <main
    class="main"
    data-testid="HOME-MAIN-TEST-ID"
  >
    <article
      class="container"
    >
      <article
        class="container"
      >
        <section
          class="imageWrapper"
        >
          <img
            alt="Clear"
            src="icons/weather-sun.svg"
          />
        </section>
        <section
          class="weatherWrapper"
          data-testid="CURRENT-TEMPERATURE-TEST-ID"
        >
          <div
            class="additionalText weatherHeader"
          >
            <span>
              Clear
            </span>
            <span>
              14° / 8°
            </span>
          </div>
          <div
            class="mainText mainTemperature"
          >
            14°
          </div>
        </section>
      </article>
      <article
        class="container"
      >
        <section>
          <div
            class="additionalText"
          >
            Altstadt
          </div>
          <section
            class="mainText date"
          >
            <div>
              Friday
            </div>
            <div>
              16. February
            </div>
          </section>
        </section>
      </article>
    </article>
    <section
      class="container"
    >
      <ol
        class="weatherItemsRow"
      >
        <li
          class="listItem"
        >
          <button
            class="container containerActive"
            type="button"
          >
            <div
              class="additionalText"
            >
              15:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              14°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              18:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              13°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              21:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              4°
            </div>
          </button>
        </li>
      </ol>
    </section>
  </main>
</div>
`;

exports[`The Home Page Component Should properly change selected on click on other weather item 2`] = `
<div>
  <main
    class="main"
    data-testid="HOME-MAIN-TEST-ID"
  >
    <article
      class="container"
    >
      <article
        class="container"
      >
        <section
          class="imageWrapper"
        >
          <img
            alt="Clear"
            src="icons/weather-sun.svg"
          />
        </section>
        <section
          class="weatherWrapper"
          data-testid="CURRENT-TEMPERATURE-TEST-ID"
        >
          <div
            class="additionalText weatherHeader"
          >
            <span>
              Clear
            </span>
            <span>
              13° / 9°
            </span>
          </div>
          <div
            class="mainText mainTemperature"
          >
            13°
          </div>
        </section>
      </article>
      <article
        class="container"
      >
        <section>
          <div
            class="additionalText"
          >
            Altstadt
          </div>
          <section
            class="mainText date"
          >
            <div>
              Friday
            </div>
            <div>
              16. February
            </div>
          </section>
        </section>
      </article>
    </article>
    <section
      class="container"
    >
      <ol
        class="weatherItemsRow"
      >
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              15:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              14°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerActive"
            type="button"
          >
            <div
              class="additionalText"
            >
              18:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              13°
            </div>
          </button>
        </li>
        <li
          class="listItem"
        >
          <button
            class="container containerInactive"
            type="button"
          >
            <div
              class="additionalText"
            >
              21:00
            </div>
            <section
              class="imageWrapper"
            >
              <img
                alt="Clear"
                src="icons/weather-sun.svg"
              />
            </section>
            <div
              class="temperature"
            >
              4°
            </div>
          </button>
        </li>
      </ol>
    </section>
  </main>
</div>
`;

exports[`The Home Page Component Should render properly if there's proper data 1`] = `
<main
  class="main"
  data-testid="HOME-MAIN-TEST-ID"
>
  <article
    class="container"
  >
    <article
      class="container"
    >
      <section
        class="imageWrapper"
      >
        <img
          alt="Clear"
          src="icons/weather-sun.svg"
        />
      </section>
      <section
        class="weatherWrapper"
        data-testid="CURRENT-TEMPERATURE-TEST-ID"
      >
        <div
          class="additionalText weatherHeader"
        >
          <span>
            Clear
          </span>
          <span>
            14° / 8°
          </span>
        </div>
        <div
          class="mainText mainTemperature"
        >
          14°
        </div>
      </section>
    </article>
    <article
      class="container"
    >
      <section>
        <div
          class="additionalText"
        >
          Altstadt
        </div>
        <section
          class="mainText date"
        >
          <div>
            Friday
          </div>
          <div>
            16. February
          </div>
        </section>
      </section>
    </article>
  </article>
  <section
    class="container"
  >
    <ol
      class="weatherItemsRow"
    >
      <li
        class="listItem"
      >
        <button
          class="container containerActive"
          type="button"
        >
          <div
            class="additionalText"
          >
            15:00
          </div>
          <section
            class="imageWrapper"
          >
            <img
              alt="Clear"
              src="icons/weather-sun.svg"
            />
          </section>
          <div
            class="temperature"
          >
            14°
          </div>
        </button>
      </li>
      <li
        class="listItem"
      >
        <button
          class="container containerInactive"
          type="button"
        >
          <div
            class="additionalText"
          >
            18:00
          </div>
          <section
            class="imageWrapper"
          >
            <img
              alt="Clear"
              src="icons/weather-sun.svg"
            />
          </section>
          <div
            class="temperature"
          >
            13°
          </div>
        </button>
      </li>
      <li
        class="listItem"
      >
        <button
          class="container containerInactive"
          type="button"
        >
          <div
            class="additionalText"
          >
            21:00
          </div>
          <section
            class="imageWrapper"
          >
            <img
              alt="Clear"
              src="icons/weather-sun.svg"
            />
          </section>
          <div
            class="temperature"
          >
            4°
          </div>
        </button>
      </li>
    </ol>
  </section>
</main>
`;

exports[`The Home Page Component Should show 404 if there's invalid data 1`] = `
<main
  class="main container"
  data-testid="ERROR-MAIN-TEST-ID"
>
  <h1>
    Well, let\`s just hope the weather will be great!
  </h1>
</main>
`;

exports[`The Home Page Component Should show 404 if there's no data 1`] = `
<main
  class="main container"
  data-testid="ERROR-MAIN-TEST-ID"
>
  <h1>
    Well, let\`s just hope the weather will be great!
  </h1>
</main>
`;
